File: demos_code_J2_conception_objet/03_isp_roles/DemoIsp.java
public class DemoIsp {
    interface PaymentReader { String find(String id); }
    interface PaymentCaptor { void capture(String id); }
    interface RefundManager { void refund(String id); }
    interface PaymentExporter { String exportCsv(); }

    static final class PaymentApplicationService
            implements PaymentReader, PaymentCaptor, RefundManager, PaymentExporter {
        public String find(String id) { return "payment:" + id; }
        public void capture(String id) { System.out.println("capture " + id); }
        public void refund(String id) { System.out.println("refund " + id); }
        public String exportCsv() { return "id,status\\nP42,CAPTURED"; }
    }

    static final class PaymentScreen {
        private final PaymentReader reader;
        PaymentScreen(PaymentReader reader) { this.reader = reader; }
        void render(String id) { System.out.println("UI -> " + reader.find(id)); }
    }

    static final class RefundUseCase {
        private final PaymentReader reader;
        private final RefundManager refunds;
        RefundUseCase(PaymentReader reader, RefundManager refunds) {
            this.reader = reader;
            this.refunds = refunds;
        }
        void execute(String id) {
            System.out.println("Avant remboursement : " + reader.find(id));
            refunds.refund(id);
        }
    }

    public static void main(String[] args) {
        PaymentApplicationService service = new PaymentApplicationService();
        new PaymentScreen(service).render("P42");
        new RefundUseCase(service, service).execute("P42");
    }
}

File: Demo en Python/demos_code_J2_conception_objet_python/03_isp_roles/demo_isp.py
from abc import ABC, abstractmethod


class PaymentReader(ABC):
    @abstractmethod
    def find(self, payment_id: str) -> str: ...


class PaymentCaptor(ABC):
    @abstractmethod
    def capture(self, payment_id: str) -> None: ...


class RefundManager(ABC):
    @abstractmethod
    def refund(self, payment_id: str) -> None: ...


class PaymentExporter(ABC):
    @abstractmethod
    def export_csv(self) -> str: ...


class PaymentApplicationService(PaymentReader, PaymentCaptor, RefundManager, PaymentExporter):
    def find(self, payment_id: str) -> str:
        return f"payment:{payment_id}"

    def capture(self, payment_id: str) -> None:
        print(f"capture {payment_id}")

    def refund(self, payment_id: str) -> None:
        print(f"refund {payment_id}")

    def export_csv(self) -> str:
        return "id,status\nP42,CAPTURED"


class PaymentScreen:
    def __init__(self, reader: PaymentReader) -> None:
        self._reader = reader

    def render(self, payment_id: str) -> None:
        print("UI -> " + self._reader.find(payment_id))


class RefundUseCase:
    def __init__(self, reader: PaymentReader, refunds: RefundManager) -> None:
        self._reader = reader
        self._refunds = refunds

    def execute(self, payment_id: str) -> None:
        print("Avant remboursement : " + self._reader.find(payment_id))
        self._refunds.refund(payment_id)


if __name__ == "__main__":
    service = PaymentApplicationService()
    PaymentScreen(service).render("P42")
    RefundUseCase(service, service).execute("P42")
